Implement RT0 in assembly

bradjc opened the pull request:
### Pull Request Overview

According to #2222 we shouldn't try to implement RAM init functions in Rust. This PR (starts to) implement rt0 in assembly for cortex-m and rv32.

I'd like to make sure we are happy with this and the names before switching boards and chips.

### Testing Strategy

in progress


### TODO or Help Wanted

Thoughts?

Also it would be great to have more eyes on my assembly.


### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

ppannuto commented on boards/hail/src/main.rs:
```diff
@@ -169,14 +169,12 @@ unsafe fn set_pin_primary_functions(peripherals: &Sam4lDefaultPeripherals) {
+/// complete.
 #[no_mangle]
-pub unsafe fn reset_handler() {
+pub unsafe fn main() {
```
```suggestion
pub unsafe fn rust_entry_point() {
```

Maybe this is a more direct name. `main` is a bit overloaded in what it means, but I believe the term "entry point" to be fairly standardized as the 'first code that is run', so calling this the "rust entry point" is perhaps a more clear term for folks reading this?

bradjc replied:
I thought main was pretty clear, since it's where all of the interesting work starts but is not the _first_ thing that runs.

`rust_entry_point` is more grepable, however.

ppannuto reviewed (commented):
I reviewed the ARM assembly pretty carefully, and it looks correct. I also double checked that all of the instructions used are valid in v6m, so this should work without modification if one of the Cortex M0-based platforms comes back into tree (interestingly, `STM <rn>!, {regs}` is valid v6m, while `STM <rn>, {regs}` is not; i.e. writeback is required)

hudson-ayers commented:
This generally looks good, though it would be nice to have some confirmation that using a naked function containing only a single inline asm block is a safe use of "rust" before memory is initialized. I only say this because I noticed that the fix to https://github.com/rust-embedded/cortex-m-rt/issues/300 used external asm, though it is possible that is just because naked_fn is unstable.

From trying to follow https://rust-lang.zulipchat.com/#narrow/stream/136281-t-lang.2Fwg-unsafe-code-guidelines/topic/The.20least.20incorrect.20init.20code.20.3A) I think that a naked inline asm function should be fine...but this is all a bit beyond me. @jonas-schievink any chance you know if this is safe? Or as safe as using external asm?

jonas-schievink commented:
I think that's fine, yes.

jonas-schievink commented:
(as safe as external asm, that is)

alistair23 reviewed (approved):
I think this looks good. I like using `main()` as I think that is what people expect to run after some setup.

lschuermann commented:
These changes look good. I'm certainly no expert on ARM assembly but the
RISC-V changes look correct to me. I think `main` is fine. A grep for
"fn main()" shouldn't yield too many results. :)

bradjc commented:
Ok I think this is now everything that needs to be changed to do this. Recent updates:

- Updated all boards to `main()` rather than `reset_handler()`.
- Changed references to `reset_handler` to `main`.
- Removed libraries/tock-rt0

bradjc commented:
@tock/core-wg This does change the boot process for all tock boards, and is probably worth a quick thought.

hudson-ayers commented:
Given the 72 files changed, I would probably recommend merging this after the 2.0 alpha 1 PR.

alevy reviewed (dismissed):
I hate that we have to do this, but it seems like we have to do this. Thanks @bradjc

alevy commented:
are we blocking on #2446 for merging this?

hudson-ayers commented:
yes I think so

bradjc commented:
Rebased and updated the new board.

ppannuto reviewed (approved):
bors r+
